[#875] add Random Limited Degree Vertex Sampling

matthiastaeschner opened the pull request:
fixes #875

merando commented on gradoop-flink/src/main/java/org/gradoop/flink/model/impl/operators/sampling/RandomLimitedDegreeVertexSampling.java:
```diff
@@ -0,0 +1,159 @@
+    DataSet<Vertex> newVertices = graph.getVertices()
+      .filter(new LimitedDegreeVertexRandomFilter<>(sampleSize, randomSeed, degreeThreshold,
+        degreeType))
+      .map(new RemoveUnnecessaryPropertiesMap<>(unnecessaryPropertyNames));
```
Please use the [PropertyRemover](https://github.com/dbs-leipzig/gradoop/blob/master/gradoop-flink/src/main/java/org/gradoop/flink/model/impl/functions/epgm/PropertyRemover.java) instead of `RemoveUnnecessaryPropertiesMap`.  Due to chaining it will be very similar in efficiency.

merando commented on gradoop-flink/src/main/java/org/gradoop/flink/model/impl/operators/sampling/RandomLimitedDegreeVertexSampling.java:
```diff
@@ -0,0 +1,159 @@
+      .equalTo(new Id<>())
+      .with(new LeftSide<>());
+
+    return graph.getConfig().getLogicalGraphFactory().fromDataSets(newVertices, newEdges);
```
You should add the GraphHeads from `graph` as well, otherwise they will be lost.

merando commented on gradoop-flink/src/main/java/org/gradoop/flink/model/impl/operators/sampling/functions/FilterVerticesWithDegreeOtherThanGiven.java:
```diff
@@ -60,14 +48,16 @@ public FilterVerticesWithDegreeOtherThanGiven(long degree) {
+    unnecessaryPropertyNames.add(VertexDegree.OUT_DEGREE_PROPERTY_NAME);
+    newVertices = newVertices
+      .filter(new VertexWithDegreeFilter<>(degree, VertexDegree.DEGREE_PROPERTY_NAME))
       .map(new RemoveUnnecessaryPropertiesMap<>(unnecessaryPropertyNames));
```
Again, also please remove the method from gradoop.

merando commented on gradoop-flink/src/main/java/org/gradoop/flink/model/impl/operators/sampling/functions/FilterVerticesWithDegreeOtherThanGiven.java:
```diff
@@ -60,14 +48,16 @@ public FilterVerticesWithDegreeOtherThanGiven(long degree) {
+      .filter(new VertexWithDegreeFilter<>(degree, VertexDegree.DEGREE_PROPERTY_NAME))
       .map(new RemoveUnnecessaryPropertiesMap<>(unnecessaryPropertyNames));
 
     graph = graph.getConfig().getLogicalGraphFactory().fromDataSets(newVertices, graph.getEdges());
```
same issue with the graph heads here

merando commented on gradoop-flink/src/main/java/org/gradoop/flink/model/impl/operators/sampling/functions/LimitedDegreeVertexRandomFilter.java:
```diff
@@ -0,0 +1,92 @@
+    long degree = 0L;
+    if (degreeType.equals(VertexDegree.DegreeType.Degree)) {
+      degree =
+        Long.parseLong(vertex.getPropertyValue(VertexDegree.DEGREE_PROPERTY_NAME).toString());
```
You can reduce this if/elseif/elseif block by adding the 'name' to the enum constant and calling degreeType.getName().

merando commented on gradoop-flink/src/main/java/org/gradoop/flink/model/impl/operators/sampling/functions/VertexDegree.java:
```diff
@@ -0,0 +1,74 @@
+    /**
+     * Input degree
+     */
+    InputDegree,
```
As mentioned before I think it would be beneficial to have something like InputDegree("_inDegree"). Obvisouly a enum constructor has to be added as well + a getName() method.

merando commented on gradoop-flink/src/main/java/org/gradoop/flink/model/impl/operators/sampling/functions/VertexDegree.java:
```diff
@@ -0,0 +1,74 @@
+   * @return An instance for DegreeType
+   */
+  public static DegreeType fromString(String degreeType) {
+    if (degreeType.equals(DegreeType.InputDegree.toString())) {
```
You can use the enum method `valueOf` to get an EnumConstant by string. If the string is not matching an IllegalArgumentException is thrown. The valueOf method is created for all enums. So if the class is written as an enum you can spare this method.

merando commented on gradoop-flink/src/main/java/org/gradoop/flink/model/impl/operators/sampling/functions/VertexDegree.java:
```diff
@@ -0,0 +1,74 @@
+/**
+ * Keeps the types of vertex-degrees and the respective property names.
+ */
+public class VertexDegree {
```
IMO the whole class should be an enum. Providing fromString, getName methods. Furthermore it would reduce the boilerplate VertexDegree.DegreeType to VertexDegree.InputDegree. 
Even though I would prefere something easier like IN, OUT, INOUT since degree is also part of the Enum or class name.

matthiastaeschner commented:
@merando Ok, thanks for your comments. I'll change that.

matthiastaeschner commented:
@merando  Changes done.

@rostam Due to the requested changes, I removed your class `RemoveUnnecessaryPropertiesMap` and adjusted the code in `RandomNonUniformVertexSampling`, also using the new ENUM `VertexDegree` now

galpha commented:
We just merged the PR fixing our licence block header. From `/**` to `/*`. Pls rebase your current branch and handle all local merge conflicts.

matthiastaeschner commented on gradoop-flink/src/test/java/org/gradoop/flink/model/impl/operators/sampling/RandomLimitedDegreeVertexSamplingTest.java:
```diff
@@ -0,0 +1,197 @@
+import static org.junit.Assert.*;
+
+@RunWith(Parameterized.class)
+public class RandomLimitedDegreeVertexSamplingTest extends GradoopFlinkTestBase {
```
There is an issue with the ParameterizedTestForGraphSampling. The current implementation won't fit for all samplings - `validateResult` and the Parameters differ depending on the sampling and the result graph (e.g. Limited Degree Vertex Sampling).
IMO we can open a new issue, refactoring and adapting the ParameterizedTestForGraphSampling - fitting for all kinds of sampling.

galpha commented:
LGTM. Like we agreed in slack pls add a issue about the validateGraph() method for **all** Sampling Methods and additional a validateSpecific() for several Sampling Methods.
